modules/shared/src/main/scala/io/constellationnetwork/schema/priceOracle.scala

```diff
@@ -0,0 +1,51 @@
+package io.constellationnetwork.schema
+
+import cats.Order._
+import cats.effect.Async
+import cats.syntax.functor._
+import cats.syntax.semigroup._
+
+import io.constellationnetwork.ext.crypto.RefinedHasher
+import io.constellationnetwork.ext.derevo.ordering
+import io.constellationnetwork.schema.artifact.PricingUpdate
+import io.constellationnetwork.schema.epoch.EpochProgress
+import io.constellationnetwork.schema.swap.CurrencyId
+import io.constellationnetwork.security.hash.Hash
+import io.constellationnetwork.security.signature.Signed
+import io.constellationnetwork.security.{Hashed, Hasher}
+
+import derevo.cats.{eqv, order, show}
+import derevo.circe.magnolia.{decoder, encoder}
+import derevo.derive
+import eu.timepit.refined.auto.autoRefineV
+import eu.timepit.refined.cats._
+import eu.timepit.refined.types.numeric.NonNegLong
+import io.estatico.newtype.macros.newtype
+
+object priceOracle {
+
+  @derive(decoder, encoder, show, order, ordering)
+  @newtype
+  case class PricingUpdateOrdinal(value: NonNegLong) {
+    def next: PricingUpdateOrdinal = PricingUpdateOrdinal(value |+| 1L)
+  }
+  object PricingUpdateOrdinal {
+    val first = PricingUpdateOrdinal(1L)
+  }
+
+  @derive(eqv, show, encoder, decoder)
+  case class PricingUpdateReference(ordinal: PricingUpdateOrdinal, hash: Hash)
+  object PricingUpdateReference {
+    val empty = PricingUpdateReference(PricingUpdateOrdinal(0L), Hash.empty)
+
+    def of(hashedTransaction: Hashed[PricingUpdate]): PricingUpdateReference =
+      PricingUpdateReference(hashedTransaction.ordinal, hashedTransaction.hash)
+
+    def of[F[_]: Async: Hasher](signedTransaction: Signed[PricingUpdate]): F[PricingUpdateReference] =
+      signedTransaction.value.hash.map(PricingUpdateReference(signedTransaction.ordinal, _))
+  }
+
+  @derive(eqv, show, encoder, decoder)
+  case class PriceRecord(currencyId: CurrencyId, fractionalValue: NonNegFraction, updatedAt: EpochProgress)
```

Comment:
Do we have a standard for how this fraction is defined? As in I previously assumed price was defined as (number of DAG per dollar) so `NonNegFraction.from(25, 1)` would mean a price of $0.04 per DAG. If we are tracking this per currencyId are we always going to reference price to USD or to DAG? Also, can we ensure that the numerator and denominator aren't flipped (maybe due to a developer misunderstanding who defines NonNegFraction.from(1, 25) since that number would be 1/25 = 0.04 but would be read as 1 dollar per 25 DAG.

Reply:
So I think datum/USD is most ideal from our discussion. But also, I think this record should be defined exclusively in terms of the latest value of the PricingUpdate and the updatedAt, since it's redundant fields otherwise (and the ref storage should also be included here so it's not in another lastref field.

So PriceRecord(latest: PricingUpdate, updatedAt: EpochProgress)